### ISSUE TITLE
parallel media forking

### ISSUE THREAD

**Issue by crddevel** [sipwise/rtpengine#318]
hi all, during ringing phase it would be great if mediaproxy-ng could fork media streams to all ringing destinations. Using kamailio and append_branch function you can call many destinations at the same time.
The only problem I see all destinations must negotiate the same media into 180 + SDP or 183 + SDP. I don't know if this is a limitation for you.

First of all, is it possible to do that?

What do you think abount that? Are you interested in this development?

If not, I would like to do from myself and it would be great if you can address me to find what I should enhanced. can you just suggest how to achieve this goal?

Thanks!

**Comment by crddevel** [comment]
Please can anyone answer my question? 
thanks

**Comment by rfuchs** [comment]
Can you give a specific example of what you would like to see? From my understanding, early media through 183 is sent from the destination to the source, so in a branched case you might see multiple early media streams going to the source, and not one from the source to possibly multiple destinations.

**Comment by crddevel** [comment]
Yes, you're right, early media is commonly used when you have for instance a SIP-to-PSTN gateway which sends back to caller the PSTN ringing tone.

The scenario I meant is the following:
- caller sends INVITE to destination groups with video and audio into SDP
- All receivers reply with 183+SDP with video support and recvonly attribute into SDP
- Once the caller receives all 183s can start sending video to all destination

Let me call it "video preview". This can be done before answering the call.

I've already done it using Asterisk and its Queue() function with some code modifications but, right now, I would like to know if this scenario can be done using kamailio + mediaproxy-ng.
So, from mediaproxy-ng side, it should fork, in parallel mode, media to multiple destinations as soon as it receives all 183s. Is it possible to fork media to multiple destinations?

Maybe, this feature/enhancement is not very commond and probably your're not interested in support it, but I would like to know if this can be done.  Do you think this enhancement is pretty hard to do, I mean I have to turn upside-down your code or not? Thanks.

**Comment by rfuchs** [comment]
I see. No, this is not currently possible and completely unsupported. A 1-to-1 mapping of input and output streams is inherent to the current design of rtpengine, so support for media forking would require significant overhaul. It's possible, but not trivial.

**Comment by crddevel** [comment]
ok, many thanks for your hints.
